Add an OpenHPC image build playbook and role

m-bull opened the pull request:
WIP while some things on ansible-slurm-appliance are still in-flight

sjpb commented:
To me, the dependencies feel the wrong way around and hard to maintain/reason about. I'd have expected an image build role to take vars for the resources (network etc) to use, and for these to already exist (if they are not stuff that the underlying packer run can create itself), and just do the image build. So, either the CI workflow would contain a step to create the required resources, or, as a matter of convenience, I'd accept it makes sense if the `build-image.yml` *playbook* calls out to both the tf role and the image build role (b/c actually, using ansible to define our CI "workflow" makes sense from the PoV of consistency and composability).

Naming nit: `image_build` role and `build-image` play seems inconsistent?

m-bull commented:
Thanks @sjpb - have refactored to make the image-build playbook call out to two roles - one to manage the infra and one to build the image. The infra management role is opt-in via `image_build_manage_infra` and I've added a pre-check to the image-build role to ensure that variables set by the now optional image_build_infra role are set in some other way. 

Also renamed the playbook to match the role.

sjpb commented on roles/image_build/templates/ansible.cfg.j2:
```diff
@@ -0,0 +1,16 @@
+[defaults]
+any_errors_fatal = True
+gathering = smart
+forks = 30
```
Only running 1x builder, could omit to get default 5 forks.

m-bull replied:
Spot - fixed

sjpb commented on roles/image_build_infra/tasks/main.yml:
Isn't this a dupe of `roles/cluster_infra`, but using different vars?

sjpb replied:
And a different resource template, but that's selectable via a var anyway ISTR.

m-bull replied:
Again, good spot - have refactored to reuse `cluster_infra`.

m-bull commented on roles/cluster_infra/defaults/main.yml:
```diff
@@ -50,3 +50,11 @@ cluster_groups_validation:
+
+# List of hw_scsi_models that result in block devices presenting as /dev/sdX
+# rather than /dev/vdX
+scsi_models:
```
I've rewritten the history to be a bit cleaner now - hopefully this fixes some of those issues.

sjpb commented on roles/image_build_infra/defaults/main.yml:
```diff
@@ -0,0 +1,12 @@
+
+# Regex to capture existing cloud image names to use as the
+# OpenHPC Slurm base-image
+image_build_existing_image_regex: "^Rocky-8-GenericCloud-Base-8.8-.*"
```
I'm not convinced that a control node built with an image based off anything after an 8.6 GenericCloud image will survive a reboot - the device labling appears to get screwed up for VMs with attached volumes.

See https://forums.rockylinux.org/t/issue-with-8-7-genericcloud-image-with-attached-volumes/8451 and https://stackhpc.slack.com/archives/C01LHHGF6BC/p1672999238711989.

m-bull replied:
Tested on a running cluster with no issues, thanks for the heads up though

sjpb commented on roles/image_build_infra/tasks/main.yml:
It'd be nicer to at least call the normal `terraform` role with some flag so that inventory is NOT created (or, make that terraform role not actually automatically run the inventory creation bit), IMO. Can live without it though!

m-bull replied:
Added here: https://github.com/stackhpc/ansible-collection-terraform/pull/7, will update in a new PR.
